src/java/org/apache/cassandra/io/sstable/format/bti/BtiTableWriter.java

```diff
@@ -0,0 +1,428 @@
+    {
+        final SequentialWriter rowIndexWriter;
+        private final FileHandle.Builder rowIndexFHBuilder;
+        private final SequentialWriter partitionIndexWriter;
+        private final FileHandle.Builder partitionIndexFHBuilder;
+        private final PartitionIndexBuilder partitionIndex;
+        boolean partitionIndexCompleted = false;
+        private DataPosition riMark;
+        private DataPosition piMark;
+
+        IndexWriter(Builder b)
+        {
+            super(b);
+            rowIndexWriter = new SequentialWriter(descriptor.fileFor(Components.ROW_INDEX), b.getIOOptions().writerOptions);
+            rowIndexFHBuilder = IndexComponent.fileBuilder(Components.ROW_INDEX, b).withMmappedRegionsCache(b.getMmappedRegionsCache());
+            partitionIndexWriter = new SequentialWriter(descriptor.fileFor(Components.PARTITION_INDEX), b.getIOOptions().writerOptions);
+            partitionIndexFHBuilder = IndexComponent.fileBuilder(Components.PARTITION_INDEX, b).withMmappedRegionsCache(b.getMmappedRegionsCache());
+            partitionIndex = new PartitionIndexBuilder(partitionIndexWriter, partitionIndexFHBuilder);
+            // register listeners to be alerted when the data files are flushed
+            partitionIndexWriter.setPostFlushListener(() -> partitionIndex.markPartitionIndexSynced(partitionIndexWriter.getLastFlushOffset()));
+            rowIndexWriter.setPostFlushListener(() -> partitionIndex.markRowIndexSynced(rowIndexWriter.getLastFlushOffset()));
+            @SuppressWarnings({"resource", "RedundantSuppression"})
+            SequentialWriter dataWriter = b.getDataWriter();
+            dataWriter.setPostFlushListener(() -> partitionIndex.markDataSynced(dataWriter.getLastFlushOffset()));
+        }
+
+        public long append(DecoratedKey key, AbstractRowIndexEntry indexEntry) throws IOException
+        {
+            bf.add(key);
+            long position;
+            if (indexEntry.isIndexed())
+            {
+                long indexStart = rowIndexWriter.position();
+                try
+                {
+                    ByteBufferUtil.writeWithShortLength(key.getKey(), rowIndexWriter);
+                    ((TrieIndexEntry) indexEntry).serialize(rowIndexWriter, rowIndexWriter.position());
+                }
+                catch (IOException e)
+                {
+                    throw new FSWriteError(e, rowIndexWriter.getFile());
+                }
+
+                if (logger.isTraceEnabled())
+                    logger.trace("wrote index entry: {} at {}", indexEntry, indexStart);
+                position = indexStart;
+            }
+            else
+            {
+                // Write data position directly in trie.
+                position = ~indexEntry.position;
+            }
+            partitionIndex.addEntry(key, position);
+            return position;
+        }
+
+        public boolean buildPartial(long dataPosition, Consumer<PartitionIndex> callWhenReady)
```

Comment:
nit: unused return value? (not a big deal, just want to make sure it's not an oversight)